### PR TITLE
KAFKA-15410: Delete topic integration test with LocalTieredStorage and TBRLMM (3/4)

### PR DESCRIPTION
**Pull request by kamalcph**
### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by kamalcph** [comment]
@clolov @divijvaidya @showuon @abhijeetk88 @satishd

This PR is stacked on top of #14307 and #14328 to review the change-set separately (we can merge them if required). Please take a look when you get chance!

File: core/src/main/java/kafka/log/remote/RemoteLogManager.java
Comment (kamalcph) [inline core/src/main/java/kafka/log/remote/RemoteLogManager.java]: @divijvaidya 

We need to revert this change (see this [comment](https://github.com/apache/kafka/pull/13947/files#r1294768292)).  In the ZK mode, controller sends two stop-partitions request while deleting the topic when the replica state changes to `Offline` and `ReplicaDeletionStarted`.  

RLMM#onStopPartitions causes the TBRLMM consumer to stop consuming records for those respective user-topic-partitions which inturn causes the segment deletion to fail for that topic (since we write a metadata record and read from that topic to denote that the metadata written successfully). 

Stop partition request will be called when a topic is deleted, partition is moved to another replica and node is stopped. In the first 2 cases, the request will have `delete` flag set to true. When the node is stopped, RLMM will be closed anyway.

File: storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerManager.java
Comment (kamalcph) [inline storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerManager.java:123]: @satishd 

Updated the logs to capture most of the information which helps to debug any issue.

File: storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerTask.java
Comment (kamalcph) [inline storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerTask.java:160]: @abhijeetk88 

This log statement will be chatty and prints the entire event. Reduced the logger level from `debug` to `trace`.

File: storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/TopicBasedRemoteLogMetadataManager.java
Comment (kamalcph) [inline storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/TopicBasedRemoteLogMetadataManager.java:380]: If the user configures the admin client incorrectly (eg: forgot to supply `remote.log.metadata.manager.listener.name`), then the thread will die without throwing any exception/log so moved the `Admin` instance creation within the try-block.

File: storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerManager.java
Comment (satishd) [inline storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerManager.java:120]: Good catch on the ordering of the offsets in the log.

**Review by satishd — COMMENTED** [review]
Thanks @kamalcph for the PR. Overall LGTM. Good to see that logs are enhanced to capture more information that will be helpful for monitoring/debugging.

File: storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/TopicBasedRemoteLogMetadataManager.java
Comment (divijvaidya) [inline storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/TopicBasedRemoteLogMetadataManager.java]: Use `Utils.closeQuietly` ?

File: core/src/main/java/kafka/log/remote/RemoteLogManager.java
Comment (divijvaidya) [reply core/src/main/java/kafka/log/remote/RemoteLogManager.java]: > Stop partition request will be called when a topic is deleted, partition is moved to another replica and node is stopped. 

The 4th case in future is going to be when TS is disabled dynamically for a topic (KIP-950). In such a scenario, TBRLMM will probably still have a problem if the disablement has a "retain=delete". We can solve it in scope for KIP-950 but heads-up that some changes in TBRLMM will be required.

> In the first 2 cases, the request will have delete flag set to true.

When replica is moving to another node, why would `deleteRemoteLog` be true? Yes, the `deleteLocalLog` will be true but not `deleteRemoteLog`. Please help me understand this.

Comment (kamalcph) [reply core/src/main/java/kafka/log/remote/RemoteLogManager.java]: yes, you're right. When the replica is moved to another node, then the RLMM#stopPartition won't be called. We can handle this case in TBRLMM by adding a test to ensure that if the same replica is reverted back to the previous node, then it gets handled gracefully. But, we need to handle this case for any RLMM implementation.

Another way to fix this issue is that we can always invoke the `RLMM#stopPartition` when `deleteLocalLog` is set to true but that requires good amount of changes.

**Comment by kamalcph** [comment]
Rebased the PR against trunk.

File: core/src/main/java/kafka/log/remote/RemoteLogManager.java
Comment (kamalcph) [reply core/src/main/java/kafka/log/remote/RemoteLogManager.java]: Addressed your review comments. Please take another look. 

As per the Javadoc in [RemoteLogMetadataManager#onStopPartitions](https://github.com/apache/kafka/blob/trunk/storage/api/src/main/java/org/apache/kafka/server/log/remote/storage/RemoteLogMetadataManager.java#L203), this method will be called when replica emigrates to a new broker and during topic deletion.

> /**
     * This method is invoked only when the topic partitions are stopped on this broker. This can happen when a
     * partition is emigrated to other broker or a partition is deleted.
     *
     * @param partitions topic partitions that have been stopped.
     */
     void onStopPartitions(Set<TopicIdPartition> partitions);

File: core/src/main/java/kafka/log/remote/RemoteLogManager.java
Comment (divijvaidya) [inline core/src/main/java/kafka/log/remote/RemoteLogManager.java:380]: else please print an warn. This is not expected.

File: core/src/main/scala/kafka/server/ReplicaManager.scala
Comment (divijvaidya) [inline core/src/main/scala/kafka/server/ReplicaManager.scala:593]: Why are we ignoring this business logic in the new revision? If you look above at the business logic of deleting the local partition, we only call delete local partition for Online partitions even if deleteLocalPartition flag is set.

File: core/src/main/java/kafka/log/remote/RemoteLogManager.java
Comment (divijvaidya) [reply core/src/main/java/kafka/log/remote/RemoteLogManager.java]: > Another way to fix this issue is that we can always invoke the RLMM#stopPartition when deleteLocalLog is set to true but that requires good amount of changes.

We want to stop archival i.e. call `leaderOrFollowerTasks.remove(tpId);` even for cases when `deleteLocalLog` is not true. I understand that you handled this in current code but wanted to call this expectation out as a reply to your above comment. 
Perhaps add a unit test (if missing) specifically to test that this expectation is not violated?

File: core/src/main/java/kafka/log/remote/RemoteLogManager.java
Comment (kamalcph) [reply core/src/main/java/kafka/log/remote/RemoteLogManager.java:380]: Added a WARN log initially, then removed it. The `stopPartition` call will come for all the partitions including the internal and non-tiered storage enabled topics. Is it ok to print in WARN level for those partitions?

File: core/src/main/java/kafka/log/remote/RemoteLogManager.java
Comment (kamalcph) [reply core/src/main/java/kafka/log/remote/RemoteLogManager.java]: Add the respective assertions in the test.

File: core/src/main/scala/kafka/server/ReplicaManager.scala
Comment (kamalcph) [reply core/src/main/scala/kafka/server/ReplicaManager.scala:593]: Reverted the logic back, I was not sure whether it will be useful to filter out the partitions with error state so ignored it.

File: core/src/main/java/kafka/log/remote/RemoteLogManager.java
Comment (divijvaidya) [reply core/src/main/java/kafka/log/remote/RemoteLogManager.java:380]: good point. So we are assuming that if a topic is in `topicIdByPartitionMap` then it has active archival otherwise not. This is not a good source of truth IMO since we are going to replace it with metadata cache in another PR (why have two caches for tp <-> tpId mapping). 

Nevertheless, let's add a comment for now here and maybe create an improvement JIRA to not call stopPArtitions for internal or non-tiered topics.

**Review by divijvaidya — APPROVED** [review]
Looks good. Thank you @kamalcph for all the work you are doing here for TS. 

I will wait for CI to pass before merging.

**Review by satishd — APPROVED** [review]
Thanks @kamalcph for addressing the review comments. LGTM.

File: storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/TopicBasedRemoteLogMetadataManager.java
Comment (showuon) [reply storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/TopicBasedRemoteLogMetadataManager.java:380]: Nice fix!

File: core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala
Comment (showuon) [inline core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala:3491]: Could you explain more about why we don't call `stopPartitions` when `throwIOException` now? My understanding is this should be a bug previously that we didn't filter out the error partition. Is that right?

**Review by showuon — APPROVED** [review]
LGTM! Just a question to clarify. Thanks.

File: core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala
Comment (kamalcph) [reply core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala:3491]: yes, there was a bug. Previously, we were calling the RLM#stopPartitions twice one with `delete` set as true and false. The error partition was by-passing through the second call.

**Comment by satishd** [comment]
Merging it to trunk and 3.6 to unblock as it received 3 +1s. A couple of test failures unrelated to this change. Merging it to trunk and 3.6.

File: core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala
Comment (showuon) [reply core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala:3491]: Good catch. Thanks.

**Pull request merged**
